Sources/SwiftLexicalLookup/LookupName.swift

```diff
@@ -135,6 +135,50 @@ import SwiftSyntax
     }
   }
 
+  /// Position of this name.
+  ///
+  /// For some syntax nodes, their position doesn't reflect
+  /// the position at which a particular name was introduced at.
+  /// Such cases are function parameters (as they can
+  /// contain two identifiers) and function declarations (where name
+  /// is precided by access modifiers and `func` keyword).
+  @_spi(Experimental) public var position: AbsolutePosition {
+    switch self {
+    case .identifier(let syntax, _):
+      return syntax.identifier.positionAfterSkippingLeadingTrivia
+    case .declaration(let syntax):
+      return syntax.name.position
+    case .implicit(let implicitName):
+      switch implicitName {
+      case .self(let declSyntax):
+        switch Syntax(declSyntax).as(SyntaxEnum.self) {
+        case .functionDecl(let functionDecl):
+          return functionDecl.name.position
+        case .initializerDecl(let initializerDecl):
+          return initializerDecl.initKeyword.positionAfterSkippingLeadingTrivia
+        case .subscriptDecl(let subscriptDecl):
+          return subscriptDecl.accessorBlock?.position ?? subscriptDecl.endPosition
+        case .variableDecl(let variableDecl):
+          return variableDecl.bindings.first?.accessorBlock?.positionAfterSkippingLeadingTrivia
+            ?? variableDecl.endPosition
+        default:
+          return declSyntax.positionAfterSkippingLeadingTrivia
+        }
+      case .Self(let declSyntax):
+        switch Syntax(declSyntax).as(SyntaxEnum.self) {
+        case .protocolDecl(let protocolDecl):
+          return protocolDecl.name.positionAfterSkippingLeadingTrivia
```

Comment:
Why do we only need this for protocols and not other type declarations?

Reply:
In this PR, protocol declarations are the only places we are introducing `Self` keyword, so handling other declarations wouldn’t matter in this case. I changed associated syntax node type of `ImplicitDecl.Self` to `ProtocolDeclSyntax`, to be more explicit and remove this nested `switch`.

Reply:
So, you’re not handling `Self` in classes yet?

```swift
class Foo {
  func bar() -> Self { fatalError() }
}
```

Reply:
Interestingly, it seems to me like `ASTScope` doesn’t introduce implicit `Self` in nominal type declarations. I suppose this could be implicitly modeled by the `lookInMembers` result? Here’s the entire output of the validation function for the snippet you’ve provided
```
-----> Lookup started at: 2:17 ("Self")
     |      ASTScope      | SwiftLexicalLookup 
> ✅ |   Look memb: 1:7   |   Look memb: 1:7   

-----> Lookup started at: 2:17 ("Self")
     |      ASTScope      | SwiftLexicalLookup 
> ✅ |   Look memb: 1:7   |   Look memb: 1:7   

-----> Lookup started at: 2:24 ("fatalError")
     |      ASTScope      | SwiftLexicalLookup 
> ✅ |      self 2:8      |      self 2:8      
> ✅ |End here: Look memb: 1:7|   Look memb: 1:7   

-----> Lookup started at: 2:24 ("fatalError")
     |      ASTScope      | SwiftLexicalLookup 
> ✅ |      self 2:8      |      self 2:8      
> ✅ |   Look memb: 1:7   |   Look memb: 1:7   
```
Nevertheless, there certainly are more places `Self` is introduced apart from protocol declarations. This seems to be also the case for extension declarations. I fixed it now as well and this is how the output looks like for the following snippet (you can also notice here the duplicate omission I’ve mentioned in my [previous comment](https://github.com/swiftlang/swift-syntax/pull/2852#issuecomment-2360959157))
```swift
protocol Foo {}
extension Foo {
  func bar() -> Self { fatalError() }
}
```
```
-----> Lookup started at: 2:11 ("Foo")
     |      ASTScope      | SwiftLexicalLookup 

-----> Lookup started at: 2:11 ("Foo")
     |      ASTScope      | SwiftLexicalLookup 

-----> Lookup started at: 3:17 ("Self")
     |      ASTScope      | SwiftLexicalLookup 
> ✅ | End here: Self 2:1 |      Self 2:1      
> ⏩ |       -----        |  Look memb: 2:11   

-----> Lookup started at: 3:24 ("fatalError")
     |      ASTScope      | SwiftLexicalLookup 
> ✅ |      self 3:8      |      self 3:8      
> ✅ |      Self 2:1      |      Self 2:1      
> ✅ |End here: Look memb: 2:11|  Look memb: 2:11   

-----> Lookup started at: 3:24 ("fatalError")
     |      ASTScope      | SwiftLexicalLookup 
> ✅ |      self 3:8      |      self 3:8      
> ✅ |      Self 2:1      |      Self 2:1      
> ✅ |  Look memb: 2:11   |  Look memb: 2:11   
> ℹ️ | Omitted ASTScope name: Self 2:1
> ℹ️ | Omitted ASTScope name: Self 2:1
```